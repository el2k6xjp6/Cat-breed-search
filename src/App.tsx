import { useEffect, useState } from "react";
import axios from "axios";
import { Breed } from "./types";
import Card from "./Card";
import SearchIcon from "./SearchIcon";
import { debounce } from "./utils";
import {
  Container,
  Header,
  InputContainer,
  Input,
  CardContainer,
} from "./styles";

function App() {
  const [queryString, setQueryString] = useState<string>("ame");
  const [breeds, setBreeds] = useState<Breed[]>([]);

  const fetchApiWithDebounce = debounce((query: string) => {
    axios
      .get("https://api.thecatapi.com/v1/breeds/search", {
        headers: {
          "x-api-key": process.env.REACT_APP_X_API_KEY as string,
        },
        params: {
          q: query,
        },
      })
      .then((res) => {
        setBreeds([...res.data]);
      });
  }, 1000);

  useEffect(() => {
    if (queryString.length >= 3) {
      fetchApiWithDebounce(queryString);
    }
  }, [queryString]);

  const handleQueryInput = (event: React.ChangeEvent<HTMLInputElement>) => {
    setQueryString(event.target.value);
  };

  return (
    <Container>
      <Header>Cat Breeds</Header>
      <InputContainer>
        <SearchIcon />
        <Input
          onChange={handleQueryInput}
          value={queryString}
          type="text"
          placeholder="The name, or part of the name."
        />
      </InputContainer>
      <CardContainer>
        {breeds.map((breed: Breed) => (
          <Card {...breed} />
        ))}
      </CardContainer>
    </Container>
  );
}

export default App;
