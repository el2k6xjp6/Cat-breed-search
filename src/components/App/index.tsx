import { useEffect, useState, useMemo } from "react";
import axios from "axios";
import { Breed } from "../../types";
import Card from "../Card";
import SearchIcon from "../SearchIcon";
import SearchBar from "../SortBar";
import { SORT_METHODS } from "../../constants";
import { debounce, getSortFunction } from "../../utils";
import {
  Container,
  Header,
  InputContainer,
  Input,
  CardContainer,
} from "./styles";

function App() {
  const [queryString, setQueryString] = useState<string>("");
  const [breeds, setBreeds] = useState<Breed[]>([]);
  const [sortMethod, setSortMethod] = useState<number>(-1);

  const fetchApiWithDebounce = useMemo(
    () =>
      debounce(
        (query: string | Breed[]) =>
          axios
            .get("https://api.thecatapi.com/v1/breeds/search", {
              headers: {
                "x-api-key": process.env.REACT_APP_X_API_KEY as string,
              },
              params: {
                q: query,
              },
            })
            .then((res) => {
              setBreeds([...res.data]);
            }),
        1000
      ),
    []
  );

  const updateResultWithDebounce = useMemo(
    () =>
      debounce((value: string | Breed[]) => {
        if (typeof value !== "string") {
          setBreeds(value);
        }
      }, 1000),
    []
  );

  useEffect(() => {
    if (queryString.length === 0) {
      updateResultWithDebounce([]);
    }
    if (queryString.length >= 3) {
      fetchApiWithDebounce(queryString);
    }
  }, [queryString, updateResultWithDebounce, fetchApiWithDebounce]);

  const handleQueryInput = (event: React.ChangeEvent<HTMLInputElement>) => {
    setQueryString(event.target.value);
  };

  const sortedBreeds = useMemo(() => {
    if (sortMethod < 0 || sortMethod >= SORT_METHODS.length) return breeds;
    const sortedBreeds = breeds.sort(getSortFunction(sortMethod));
    if (sortMethod % 2 === 1) {
      sortedBreeds.reverse();
    }
    return sortedBreeds;
  }, [breeds, sortMethod]);

  return (
    <Container>
      <Header>Cat Breeds</Header>
      <InputContainer>
        <SearchIcon />
        <Input
          onChange={handleQueryInput}
          value={queryString}
          type="text"
          placeholder="The name, or part of the name."
        />
      </InputContainer>
      {queryString.length !== 0 && (
        <SearchBar selection={sortMethod} handleClick={setSortMethod} />
      )}
      <CardContainer>
        {sortedBreeds.map((breed: Breed) => (
          <Card {...breed} />
        ))}
      </CardContainer>
    </Container>
  );
}

export default App;
