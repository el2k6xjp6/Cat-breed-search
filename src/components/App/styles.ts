import styled from "styled-components";

export const Container = styled.div`
  width: 90%;
  margin: 0 5%;
`;

export const Header = styled.h1`
  text-align: center;
`;

export const CardContainer = styled.div`
  margin-top: 20px;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
`;
