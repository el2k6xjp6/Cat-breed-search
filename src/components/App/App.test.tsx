import App from ".";
import "jest-styled-components";
import renderer from "react-test-renderer";

describe("App", () => {
  test("snapshot for App", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
