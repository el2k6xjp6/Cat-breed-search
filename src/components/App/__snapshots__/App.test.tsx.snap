// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot 1`] = `
.c2 {
  border: 1px solid #5f6368;
  border-radius: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  place-items: center;
  width: 60%;
  height: 40px;
  margin: 0 20%;
}

.c4 {
  margin: 0 auto;
  border: none;
  border-radius: 24px;
  height: 38px;
  width: 95%;
  font-size: 18px;
}

.c4:focus {
  outline-width: 0;
}

.c3 {
  height: 30px;
  width: 30px;
  margin-left: 10px;
}

.c6 {
  cursor: pointer;
  width: 200px;
  height: 40px;
}

.c7 {
  position: relative;
  border-radius: 4px;
  border: solid 1px #c4c4c4;
  height: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: white;
}

.c7:hover {
  border: solid 1px black;
}

.c8 {
  text-indent: 15px;
}

.c9 {
  color: #aaaaaa;
}

.c10 {
  position: absolute;
  right: 10px;
  color: #909090;
  -webkit-transform: rotate(360deg);
  -ms-transform: rotate(360deg);
  transform: rotate(360deg);
  -webkit-transition: -webkit-transform .2s ease-in-out;
  -webkit-transition: transform .2s ease-in-out;
  transition: transform .2s ease-in-out;
}

.c5 {
  width: 100%;
  height: 60px;
  margin: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 {
  width: 90%;
  margin: 0 5%;
}

.c1 {
  text-align: center;
}

.c11 {
  margin-top: 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

<div
  className="c0"
>
  <h1
    className="c1"
  >
    Cat Breeds
  </h1>
  <div
    className="c2"
  >
    <svg
      className="c3"
    >
      <circle
        cx="12"
        cy="12"
        fill="white"
        r="8"
        stroke="black"
        strokeWidth="3"
      />
      <line
        stroke="black"
        strokeWidth="3px"
        x1="18"
        x2="26"
        y1="18"
        y2="26"
      />
    </svg>
    <input
      className="c4"
      id="input"
      onChange={[Function]}
      placeholder="The name, or part of the name."
      type="text"
      value="ame"
    />
  </div>
  <div
    className="c5"
  >
    <div
      className="c6"
    >
      <div
        className="c7"
        id="DropDownHeader"
        onClick={[Function]}
      >
        <div
          className="c8 c9"
        >
          Sort By
        </div>
        <div
          className="c10"
          id="DropDownArrow"
        >
          â–¼
        </div>
      </div>
    </div>
  </div>
  <div
    className="c11"
  />
</div>
`;
